src/main/java/pl/jagiellonian/Parsers/ExpressionParser.java

```diff
@@ -0,0 +1,77 @@
+package pl.jagiellonian.Parsers;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ExpressionParser {
+    private static final Pattern MULTIPLE_EXPRESSION = Pattern.compile("((\\*?)x_(?:[1-9][0-9]*(?:\\^-?[0-9]+)*))*");
+    private static final Pattern CONSTANT_IN_EXPRESSION = Pattern.compile("\\(?(-?[1-9][0-9]*)\\)?(.*)");
+    private static final Pattern VARIABLE_IN_EXPRESSION = Pattern.compile("x_([1-9][0-9]*(?:\\^-?[0-9]+)*)");
+
+    private ExpressionParser() {
+    }
+
+    public static ParsedExpression parseExpression(String expression) {
+        ParsedExpression parsedExpression = new ParsedExpression();
+        Matcher constantMatcher = CONSTANT_IN_EXPRESSION.matcher(expression);
+        if (constantMatcher.matches()) {
+            parsedExpression.setConstant(Integer.parseInt(constantMatcher.group(1)));
+        }
+        Matcher variableMatcher = VARIABLE_IN_EXPRESSION.matcher(expression);
+        while (variableMatcher.find()) {
+            String variable = variableMatcher.group(1);
+            String[] split = variable.split("\\^");
+            int index = Integer.parseInt(split[0]);
+            int power = parsedExpression.getVariablePower(index);
+            if (split.length > 1) {
+                power += Integer.valueOf(split[1]);
```

Comment:
Boxing/unboxing to parse a primitive
A boxed primitive is created from a String, just to extract the unboxed primitive value. It is more efficient to just call the static parseXXX method.